compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+// for comparison equality
+func (c *CompressionVerifier) GetHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	return c.decompressAndHash(table, columns, rows, tableCompression)
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
+	quotedPK := quoteField(pkColumn)
+	sql, args, err := rowSelector(columns, pkColumn).
+		From(QuotedTableNameFromString(schema, table)).
+		Where(sq.Eq{quotedPK: pks}).
+		OrderBy(quotedPK).
+		ToSql()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// This query must be a prepared query. If it is not, querying will use
+	// MySQL's plain text interface, which will scan all values into []uint8
+	// if we give it []interface{}.
+	stmt, err := db.Prepare(sql)
+	if err != nil {
+		return nil, err
+	}
+
+	defer stmt.Close()
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+func (c *CompressionVerifier) decompressAndHash(table string, columns []schema.TableColumn, rows *sql.Rows, tableCompression map[string]string) (map[uint64][]byte, error) {
+	defer rows.Close()
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanGenericRow(rows, 2)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := rowData.GetUint64(0)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decodedRowData contains a map of all
+		// columns and associated decoded values by the index of the column
+		var decodedRowData map[int][]byte
+		for idx, column := range columns {
+			// Check if column is configured as compressed and decompress if necessary
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				decodedColData, err := c.Decode(table, column.Name, algorithm, rowData[idx].([]byte))
+				if err != nil {
+					return nil, err
+				}
+				decodedRowData[idx] = decodedColData
+			} else {
+				// Do not compress the column, add it to the decodedRowData to be fingerprinted
+				decodedRowData[idx] = rowData[idx].([]byte)
+			}
+		}
+
+		// Hash the data of the row to be added to the result set
+		decodedRowHash, err := c.HashRow(decodedRowData)
+		if err != nil {
+			return nil, err
+		}
+
+		resultSet[pk] = decodedRowHash
+	}
+
+	return resultSet, nil
+}
+
+// Decode will apply the configured decompression algorithm to the configured columns data
+func (c *CompressionVerifier) Decode(table, column, algorithm string, compressed []byte) ([]byte, error) {
```

Comment:
Same here, for clarity's sake, can we call it `Decompress` instead of `Decode`?

Reply:
Changed, but yeah it was definitely the `Snappy` library that threw me off and caused me to start naming things using `decode` instead of just `decompress`.